cocos/base/CCGLUtils.cpp

```diff
@@ -24,151 +24,204 @@
 
 #include "CCGLUtils.h"
 #include <stdio.h>
+#include <cfloat>
+#include <cassert>
 
 NS_CC_BEGIN
 
+// todo: use gl to get the supported number
+#define MAX_ATTRIBUTE_UNIT  16
+
 //FIXME: Consider to use variable to enable/disable cache state since using macro will not be able to close it if there're serious bugs.
 // #undef CC_ENABLE_GL_STATE_CACHE
 // #define CC_ENABLE_GL_STATE_CACHE 0
 
 #if CC_ENABLE_GL_STATE_CACHE
-
-namespace {
-    const int MAX_ACTIVE_TEXTURE = 16;
-    GLuint __currentArrayBufferId = -1;
-    GLuint __currentElementArrayBufferId = -1;
-    GLuint __activeTextureUnit = 0;
-    GLuint __currentBoundTexture[MAX_ACTIVE_TEXTURE] =  {
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1,
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1,
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1,
-        (GLuint)-1,(GLuint)-1,(GLuint)-1,(GLuint)-1
-    };
+namespace
+{
+    GLint __currentVertexBuffer = -1;
+    GLint __currentIndexBuffer = -1;
+    GLint __currentVertexArray = -1;
+    
+    uint32_t __enabledVertexAttribArrayFlag = 0;
+    VertexAttributePointerInfo __enabledVertexAttribArrayInfo[MAX_ATTRIBUTE_UNIT];
 }
-
 #endif // CC_ENABLE_GL_STATE_CACHE
 
 //FIXME: need to consider invoking this after restarting game.
 void ccInvalidateStateCache()
 {
 #if CC_ENABLE_GL_STATE_CACHE
-    __currentArrayBufferId = -1;
-    __currentElementArrayBufferId = -1;
-    __activeTextureUnit = 0;
-    for (int i = 0; i < MAX_ACTIVE_TEXTURE; ++i)
-    {
-        __currentBoundTexture[i] = -1;
-    };
+    __currentVertexBuffer = -1;
+    __currentIndexBuffer = -1;
+    __currentVertexArray = -1;
+    
+    __enabledVertexAttribArrayFlag = 0;
+    for (int i = 0; i < MAX_ATTRIBUTE_UNIT; ++i)
+        __enabledVertexAttribArrayInfo[i] = VertexAttributePointerInfo();
```

Comment:
It's unneeded to initialize __enabledVertexAttribArrayInfo array since all elements in it will be initialized by VertexAttributePointerInfo constructor.